### ISSUE TITLE
Line blocks

### ISSUE THREAD

**Issue by jgm** [jgm/djot#29]
Pandoc offers a line block format which is useful for addresses and such things as poetry.  In this format, newlines are hard breaks, and all spaces are significant, even leading spaces.
```
| The limerick packs laughs anatomical
| Into space that is quite economical.
|   But the good ones I've seen
|   So seldom are clean
| And the clean ones so seldom are comical.
```

**Comment by uvtc** [comment]
I didn't realize you could use backslash-space to make consecutive spaces that would not be condensed down to a single space. (I've previously used something like `\hspace{0.5cm}` for that (when targetting pdf)).

**Comment by matklad** [comment]
I think this can be generalized to a block, which applies only _inline_ formatting, ignores block formatting and keeps whitespace. That is, verbatim + inlines. 

This potentially could be useful for https://github.com/jgm/djot/discussions/49.

**Comment by Omikhleia** [comment]
> Note that the same effect can be achieved with backslash-newline and backslash-space, but it arguably looks less natural.

Just my two cents regarding the semantics: what does that backlash-space _really_ means here actually?

In regular justified content, it's usually a non-breaking space, so typically a U+00A0... But in proper typography those are stretchable and shrinkable as any interword-space, i.e. have a variable width. See  [Unicode TR14](https://unicode.org/reports/tr14/) §3.

In regular poetry, I'd say that the initial "indent" of a verse is (usually) a fixed-width space. It _could_ be, for instance an em-quad (U+2001)... Whatever, it's usually not adjustable.

To work around this, a rendering engine could say, well, that `\ ` may be seen as stretchable _except_ when it's initial in a paragraph or after a force line break, but that seems a bit contrived, perhaps?[^1]

I am not saying that Djot shall necessarily have Pandoc-line Line Blocks --- just that the suggested "same effect" using existing syntax might not be fully equivalent, and might impose a mixed interpretation of backslash-space depending on context.

[^1]: For the record, that's how I interpret the initial space in Pandoc-like Line Blocks in [markdown.sile](https://github.com/Omikhleia/markdown.sile) in the context of poetry typesetting.

**Comment by dereuromark** [comment]
I actually implemented it as such:
https://php-collective.github.io/djot-php/guide/syntax.html#line-blocks

The main problem with pipes is that it is harder to work with without smart IDE (rearranging your hard breaks is cumbersome).

I want to share another idea I am currently exploring:

A mechanism that sidesteps both reservations raised in the opening post: reuse the generic div instead of a new `|` line prefix.

```
::: verse
The limerick packs laughs anatomical
Into space that is quite economical.
  But the good ones I've seen
  So seldom are clean
And the clean ones so seldom are comical.
:::
```

Inside a `verse` div, soft line breaks are preserved (a single newline renders as `<br>`), and leading spaces are kept. Everything else parses as normal inline djot, so emphasis, links, and so on still work.

Alternative names:
- literal (maybe a bit more agnostic)
- strict
- line-block

### Why this avoids the two objections

1. **No pipe-table confusion.** The original concern was that a leading `|` collides with pipe tables. The div form uses no `|` at all, so there is nothing to disambiguate. The opening `:::` fence is already unambiguous.
2. **"Not sure it's worth it."** As a div behavior it costs nothing for anyone who never writes `::: verse`: it is opt-in, needs no new block syntax, and reuses machinery that already exists. The backslash-newline workaround stays available; this just removes the per-line noise for the cases (poetry, addresses, lyrics, signature blocks) where every line would otherwise need a `\`.

### Semantics

- A single newline inside the div becomes a hard break (`<br>` in HTML; a literal newline in plain-text output).
- Leading whitespace on a line is significant (preserved), so stanza indentation survives.
- A blank line still separates stanzas / paragraphs within the verse, as elsewhere.
- The div is the unit of "preserve breaks"; outside it, newlines behave normally.

### Open questions (same shape as any reserved-div feature)

- Is `verse` a reserved div class with built-in rendering, or a name a processor opts into? Either works; the reserved-class route keeps output identical across implementations.
- How far should leading-whitespace preservation go (raw spaces vs. a normalized indent)?
- Interaction with a caption/attribution line for the poet or source.


Not pushing for it, just offering the mechanism: if line blocks ever land in djot, the `::: verse` form may be the lower-friction path than a `|` prefix.

I personally find it nice, because then I can drop in "Markdown" formatted text and it just works within those divs and formats the newlines as it would be in other MD docs.

**Comment by jgm** [comment]
I don't like the idea of an attribute on generic divs that changes how the contents of the div are *parsed*.
However, we could *almost* leave parsing alone and let the *renderer* handle the `verse` attribute ; after all, soft breaks are included in the AST and can be rendered as hard breaks in this context.  The leading indentation could not be handled this way, though, as this disappears in the AST.

I also don't like the idea of relying on English-language keywords in the syntax, as noted before.

**Comment by dereuromark** [comment]
Well, you could also just use the pipe here ;)
Given thats the symbol you worked with so far in the idea:

```
::: |
The limerick packs laughs anatomical
Into space that is quite economical.
  But the good ones I've seen
  So seldom are clean
And the clean ones so seldom are comical.
:::
```

It would just replace the requirement of having to prefix every row with the | itself, which is hard to work with.
It also works for nested content - I tested It.

And it would internally replace it with `line-block` class or alike for rendering then.

POC: https://github.com/php-collective/djot-php/pull/243
Keeping this an optional extension until specs include it.

I deployed a real-live demo for us [here](https://sandbox.dereuromark.de/sandbox/djot/extensions#ext-line_block_div)

**Comment by Omikhleia** [comment]
> How far should leading-whitespace preservation go (raw spaces vs. a normalized indent)?

That's a good question.

_Personally_ I never really liked that feature (from the markdown line blocks extension), which is a bit unclear (is that supposed to expand to an em-quad per actual space?)... and incomplete: for _verses_, at least, it's also frequent in some works to mark the middle of a verse with a longer space (a.k.a. medial gap, extended spacing, caesural spacing &c.)
It seems a bit weird to have some special handling for the initial spaces, and not elsewhere.

<img width="389" height="109" alt="Image" src="https://github.com/user-attachments/assets/e1ec4057-c265-4acd-ae26-ec03418fa54a" />

**Comment by dereuromark** [comment]
I agree, if we go "line-block" it probably should preserve any and all extra indentation, which includes "inline alignment".
That's the consistent position, and it matches Pandoc's own line-block definition ("all spaces are significant"). If the construct's whole purpose is significant whitespace, preserving only the leading run while collapsing medial gaps is half a feature.

//EDIT
https://github.com/php-collective/djot-php/pull/244
Updated the live demo above accordingly.

**Comment by dereuromark** [comment]
I also had the idea of adding also `::: \` divs as a lightweight version (only applying the markdown behavior of auto newlines):

```
::: \
Some  awesome
 text
:::
```

Results in what one would expect:
```
Some awesome
text
```

no alignment or leading significant space

Both solutions are only primary level, though, they dont inherit it to sub elements.
Thats why djot-php version has a "Break mode" toggle from SoftBreak to HardBreak also as a global lever.
